### PR TITLE
Use agent instead of atom to implement mock connection

### PR DESCRIPTION
**Pull request by ChrisBlom**
Uses an agent to manage the state of the mock connection, 
see https://github.com/vvvvalvalval/datomock/issues/2

### CODE REVIEW

**Comment by vvvvalvalval** [comment]
Can't merge this right now because it may leave some concurrency issues unaddressed. In particular, the proposed change does not let you read your own writes with `(datomic.api/db conn)`, and I'm not sure that's okay (stay tuned to [this StackOverflow question](https://stackoverflow.com/questions/47693495/datomic-on-a-peer-does-connection-db-read-your-writes-after-connection-trans)). 

Even if it *is* okay, the need for a proper implementation of `sync` will probably become more necessary.

I'll keep you posted as I make progress.

**Comment by ChrisBlom** [comment]
According to the docs `transact` returns a *completed* future, so that would suggest that 
after `(d/transact conn data)` finished, the data is transacted, and your writes should be in  `(d/db conn)` , but it would be good if someone from Datomic can confirm this. 

I see now that my implementation is not correct, i've implemened `transact-async` not `transact`, 
i'll fix this. I this the concurrency issue you mean?

**Comment by ChrisBlom** [comment]
I noticed another problem: the future is delivered before the fn that updates the agent returns, so 
potentially the agent state is not updated yet once the future is delivered, do you now a clean way to solve this?

**Comment by vvvvalvalval** [comment]
> I noticed another problem: the future is delivered before the fn that updates the agent returns, so
potentially the agent state is not updated yet once the future is delivered, do you now a clean way to solve this?

@ChrisBlom it is not another problem, it is the same problem :)

**Comment by vvvvalvalval** [comment]
@ChrisBlom thanks, I'd appreciate if you could add a little test for the RYW semantics

**Comment by vvvvalvalval** [comment]
And we still need a proper implementation of `(d/sync conn t)`, if you want to give it a shot the best strategy I can think of is adding a watch for to the agent that looks up the basis-t (then self-removes)

File: src/datomock/impl.clj [hunk 1]
```diff
@@ -83,4 +86,8 @@
 
 (defn mock-conn*
   [^Database db, ^Log parent-log]
-  (->MockConnection (atom (->MockConnState db [])) (d/next-t db) parent-log (atom nil)))
+  (let [state-agent (-> (agent (->MockConnState db [] nil))
+                        (add-watch ::force-deliver-promise (fn [_ _ old new]
+                                                             (when-not (= old new)
```
Comment (ChrisBlom) [inline src/datomock/impl.clj]: should use `identical?` here

Comment (vvvvalvalval) [reply src/datomock/impl.clj]: @ChrisBlom yes, it will be equivalent and faster since the `deliver-tx-res` field will not be collection-typed.

**Comment by ChrisBlom** [comment]
I've implemented `(d/sync conn t)` and added a test for `d/transact-async`

**Comment by ChrisBlom** [comment]
the implementation is not correct, `(d/sync conn t)` does not work when the db already has t, please ignore for now

**Pull request closed without merging**